pkg/http/request_log.go

```diff
@@ -0,0 +1,164 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"text/template"
+	"time"
+)
+
+// RequestLogHandler implements an http.Handler that writes request logs
+// and calls the next handler.
+type RequestLogHandler struct {
+	handler     http.Handler
+	inputGetter RequestLogTemplateInputGetter
+	writer      io.Writer
+	templateMux sync.RWMutex
+	templateStr string
+	template    *template.Template
+}
+
+// RequestLogRevision provides revision related static information
+// for the template execution.
+type RequestLogRevision struct {
+	Name          string
+	Namespace     string
+	Service       string
+	Configuration string
+	PodName       string
+	PodIP         string
+}
+
+// RequestLogResponse provided response related information for the template execution.
+type RequestLogResponse struct {
+	Code    int
+	Size    int
+	Latency float64
+}
+
+// RequestLogTemplateInput is the wrapper struct that provides all
+// necessary information for the template execution.
+type RequestLogTemplateInput struct {
+	Request  *http.Request
+	Response *RequestLogResponse
+	Revision *RequestLogRevision
+}
+
+// RequestLogTemplateInputGetter defines a function returning the input to pass to a request log writer.
+type RequestLogTemplateInputGetter func(req *http.Request, resp *RequestLogResponse) *RequestLogTemplateInput
+
+// RequestLogTemplateInputGetterFromRevision returns a func that forms a template input using a static
+// revision information.
+func RequestLogTemplateInputGetterFromRevision(rev *RequestLogRevision) RequestLogTemplateInputGetter {
+	return func(req *http.Request, resp *RequestLogResponse) *RequestLogTemplateInput {
+		return &RequestLogTemplateInput{
+			Request:  req,
+			Response: resp,
+			Revision: rev,
+		}
+	}
+}
+
+// NewRequestLogHandler creates an http.Handler that logs request logs to an io.Writer.
+func NewRequestLogHandler(h http.Handler, w io.Writer, templateStr string,
+	inputGetter RequestLogTemplateInputGetter) (*RequestLogHandler, error) {
+	reqHandler := &RequestLogHandler{
+		handler:     h,
+		writer:      w,
+		inputGetter: inputGetter,
+	}
+	if err := reqHandler.SetTemplate(templateStr); err != nil {
+		return nil, err
+	}
+	return reqHandler, nil
+}
+
+// SetTemplate sets the template to use for formatting request logs.
+// Setting the template to an empty string turns of writing request logs.
+func (h *RequestLogHandler) SetTemplate(templateStr string) error {
+	var t *template.Template
+	// If templateStr is empty, we will set the template to nil
+	// and effectively disable request logs.
+	if templateStr != "" {
+		// Make sure that the template ends with a newline. Otherwise,
+		// logging backends will not be able to parse entries separately.
```

Comment:
Should we check this after applying the template?  I bet it would be more reliable?

Reply:
No sure if I understand - if we check this after we apply, compiled template won't have the newline and will not work.

Reply:
I see, because it executes it directly to a writer...